File: docs/dp_tabulation/1872_hard/merge_stone_game.cpp
#include <vector>
#include <numeric>
using namespace std;

int maximizeScoresDifference(vector<int> &stones) // LeetCode Q.1872.
{
    int prefixSum = 0;
    for (const auto &stone : stones)
        prefixSum += stone;

    // Base case: Alice merges everything as the only attempt.
    int maxScoreDiff = prefixSum;

    prefixSum -= stones.back();

    for (int idx = stones.size() - 2; idx >= 1; idx--)
    {
        if (prefixSum - maxScoreDiff > maxScoreDiff)
            maxScoreDiff = prefixSum - maxScoreDiff;

        prefixSum -= stones[idx];
    }

    return maxScoreDiff;
}

File: docs/dp_tabulation/1872_hard/merge_stone_game.py
def maximize_scores_difference(stones: list[int]) -> int:
    prefix_sum = sum(stones)

    # Base case: Alice merges everything as the only attempt.
    max_score_diff = prefix_sum

    prefix_sum -= stones[-1]

    for idx in range(len(stones) - 2, 0, -1):
        if prefix_sum - max_score_diff > max_score_diff:
            max_score_diff = prefix_sum - max_score_diff

        prefix_sum -= stones[idx]

    return max_score_diff
